rllib/algorithms/algorithm.py

```diff
@@ -1316,12 +1316,27 @@ def restore_workers(self, workers: WorkerSet):
         restored = workers.probe_unhealthy_workers()
 
         if restored:
-            from_worker = workers.local_worker() or self.workers.local_worker()
-            state = ray.put(from_worker.get_state())
+            # Figure out whether we are restoring a worker from the eval worker set.
+            is_eval_worker_set = False
+            from_worker = workers.local_worker()
+            if from_worker is None:
+                from_worker = self.workers.local_worker()
+                is_eval_worker_set = workers is self.evaluation_workers
+
+            # Get the state of the correct (reference) worker. E.g. The local worker
+            # of the main WorkerSet.
+            state = from_worker.get_state()
+            # For the evaluation set, we need to adjust the `policy_mapping_fn` and
+            # `is_policy_to_train` fn from the original evaluation config.
+            if is_eval_worker_set:
+                del state["policy_mapping_fn"]
+                del state["is_policy_to_train"]
```

Comment:
what if we simply always del these 2 fields? you then don't have to guess whether we are dealing with an eval worker or not.
given how worker actors are restored, they should have proper mapping_fn and policy_to_train upon recovery?

Reply:
That would break the case where you have already changed the `policy_mapping_fn` a couple of times (it's different now than the original one from your config; which is why these are part of the state to begin with) and are now trying to re-start a crashed remote worker and make sure it uses the correct mapping fn.